extensions/basic_auth/plugin.cc

```diff
@@ -0,0 +1,330 @@
+  // basic_auth_configuration_ container has the following example structure
+  //{
+  // "exact":{
+  //    "/products/api/store":{
+  //      ["GET"],
+  //      ["YWRtaW46YWRtaW4="]
+  //    }
+  // },
+  // "prefix":{
+  //    "/reviews":{
+  //      ["GET", "POST"],
+  //      ["YWRtaW46YWRtaW4=", "AWRtaW46YWRtaW4="]
+  //    }
+  //  },
+  // "suffix":{
+  //    "/api/pay":{
+  //      ["GET", "POST"],
+  //      ["YWRtaW46YWRtaW4=", "AWRtaW46YWRtaW4="]
+  //    }
+  //  }
+  //}
+  if (!JsonArrayIterate(
+          j, "basic_auth_rules", [&](const json& configuration) -> bool {
+            auto request_path =
+                JsonGetField<std::string>(configuration, "request_path")
+                    .value_or("");
+            auto match =
+                JsonGetField<std::string>(configuration, "match").value_or("");
+            if (request_path == "" || match == "") {
+              return false;
+            }
+            struct headerData header;
+            if (!JsonArrayIterate(
+                    configuration, "request_methods",
+                    [&](const json& methods) -> bool {
+                      auto method = JsonValueAs<std::string>(methods);
+                      if (method.second !=
+                          Wasm::Common::JsonParserResultDetail::OK) {
+                        LOG_WARN("unexpected method");
+                        return false;
+                      }
+                      header.request_methods.insert(
+                          std::string(method.first.value()));
+                      return true;
+                    })) {
+              LOG_WARN("Failed to parse configuration for request methods.");
+              return false;
+            }
+
+            if (!JsonArrayIterate(
+                    configuration, "credentials",
+                    [&](const json& credentials) -> bool {
+                      auto credential = JsonValueAs<std::string>(credentials);
+                      if (credential.second !=
+                          Wasm::Common::JsonParserResultDetail::OK) {
+                        LOG_WARN("unexpected credential");
+                        return false;
+                      }
+                      header.encoded_credentials.insert(
+                          Base64::encode(credential.first.value().data(),
+                                         credential.first.value().size()));
+                      return true;
+                    })) {
+              LOG_WARN("Failed to parse configuration for credentials.");
+              return false;
+            }
+            auto path_iter = basic_auth_configuration_.find(match);
+            if (path_iter != basic_auth_configuration_.end() &&
+                path_iter->second.find(request_path) !=
+                    basic_auth_configuration_[match].end()) {
+              for (auto& methods : header.request_methods) {
+                basic_auth_configuration_[match][request_path]
```

Comment:
This is a bit strange layout. I would expect it to be keyed by request method since there is a fixed limited set of them. Paths are usually prefix-matched, which is difficult to do without iterating over the map.

Reply:
I chose `match` to be the key because first I try to find an exact match for the requested path, if that's not the case I proceed to iterate over the request_paths that are mapped to prefix and suffix match patterns in order to find a prefix or suffix that matches the HTTP request path. I don't know if this makes sense. I am basing my logic off of this example: https://swagger.io/docs/specification/2-0/authentication/basic-authentication/ .If I change the key to be `request_method` instead of `match` I would still need to have as key of my second map `match`, right?

Reply:
I agree with Kuat on this. Using pattern as key is not saving you any runtime on searching. It makes more sense to use method as key and value could just be an array of struct, which has a path matcher{prefix, exact, suffix} and allowed credentials on each matcher.